### PR TITLE
Data cleaning

### LINKED ISSUE #99: Data cleaning
I recently saw that our data needs a little cleaning. For example, same licenses are called differently here:
- https://github.com/NFDI4BIOIMAGE/training/blob/edf1e0ded6b77ab4ff8ed04c074cf955e638419a/resources/materials.yml#L16
- https://github.com/NFDI4BIOIMAGE/training/blob/edf1e0ded6b77ab4ff8ed04c074cf955e638419a/resources/materials.yml#L51
- https://github.com/NFDI4BIOIMAGE/training/blob/edf1e0ded6b77ab4ff8ed04c074cf955e638419a/resources/materials.yml#L1515

We should also check other entries such as tags if there might be similar issues.

It would be great if we could standardize these meta data entries, of course ideally automatically. The notebook here might help as it summarizes existing licenses:

https://github.com/NFDI4BIOIMAGE/training/blob/main/scripts/license_statistics.ipynb

@SeverusYixin would you mind having looking into this? Thank you!

### PR DESCRIPTION
**Pull request by SeverusYixin**
This script normalizes variables in YAML files, particularly focusing on the following:

License Names:

It fetches the SPDX license list from the SPDX website and uses it to standardize license names to their canonical SPDX IDs.

Author Names:

It handles different formats of author names, including:
"Last, First" format.
"First Last" format.
Complex structures like "Last, First, Last, First".
Ensures consistency in author name representation across the data.

Tags:

Normalizes tags to a consistent format, ensuring they are uniformly capitalized and stripped of extraneous spaces.

Types:

Ensures all types are represented as lists and are uniformly capitalized and stripped of extraneous space.

Hi @haesleinhuepf, would you mind to review this PR when you have a moment. Thanks a lot :)

This PR closes #99

### CODE REVIEW

**Comment by haesleinhuepf** [comment]
Hi @SeverusYixin ,

I'm happy to review this. Would you mind writing some sentences what these changes are doing? What features did you implement?

Thanks!

Best,
Robert

**Comment by SeverusYixin** [comment]
> Hi @SeverusYixin ,
> 
> I'm happy to review this. Would you mind writing some sentences what these changes are doing? What features did you implement?
> 
> Thanks!
> 
> Best, Robert

Let me check it

**Review by haesleinhuepf — CHANGES_REQUESTED** [review]
Hi @SeverusYixin ,

thanks for sending this PR! You find some detailed comments below. Furthemore, I was hoping that this code would be execute as part of the [github workflows](https://github.com/NFDI4BIOIMAGE/training/tree/main/.github/workflows). Yould you please enter it there and make sure that the scripts are executed when a pull-request is sent (test.yml) or when the book is deployed (book.yml).

Please change those things and tag me afterwards, I'm happy to take another look.

Thanks!

Best,
Robert

File: scripts/data_normalizer.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+# Normalize author names
+def normalize_author_name(name):
+    # name = name.replace('\xE1', 'á').replace('\xE9', 'é').replace('\xED', 'í').replace('\xF3', 'ó').replace('\xFA', 'ú')  # Handling special characters
+    parts = [part.strip() for part in name.split(',')]
+    if len(parts) == 2:
+        return f"{parts[1]} {parts[0]}"
+    return name
+
+# Normalize a list of authors
+def normalize_author_list(authors):
```
[the remark this answers is not in the document]
Comment (SeverusYixin) [reply scripts/data_normalizer.py:55]: It should be the code that I had blocked out earlier that I inadvertently unblocked, I've removed and updated it

File: scripts/data_normalizer.py [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+                # Handle Type 1: "lastname, firstname"
+                for subpart in subparts:
+                    normalized_authors.append(normalize_author_name(subpart.strip()))
+        else:
+            # Handle Type 2: "firstname lastname"
+            normalized_authors.append(author)
+
+    return normalized_authors
+
+def normalize_author_list(authors):
```
[the remark this answers is not in the document]
Comment (SeverusYixin) [reply scripts/data_normalizer.py]: the first one, I have update it

File: scripts/data_normalizer.py [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+
+# Specifically handle normalization of 'type' to ensure all outputs are lists
+def normalize_type(type):
+    if isinstance(type, list):
+        return [type_.strip().title() for type_ in type]
+    else:
+        return [type.strip().title()]
+
+# Create a mapping for items (authors, tags) to a consistent format
+def create_mapping(items):
```
[the remark this answers is not in the document]
Comment (SeverusYixin) [reply scripts/data_normalizer.py:40]: Do you want to do this next time or do you want me to change the current code, I'm still working on it

**Comment by SeverusYixin** [comment]
@haesleinhuepf Would you mind to review it again

File: .github/workflows/book.yml [hunk 4]
```diff
@@ -32,6 +32,11 @@ jobs:
         python scripts/export_csv.py
         jupyter-book build docs/
 
+    # Run data normalizer script after builded the book
```
Comment (haesleinhuepf) [inline .github/workflows/book.yml:35]: Shouldn't the data normalizer be executed before building the book?

File: scripts/data_normalizer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (haesleinhuepf) [reply scripts/data_normalizer.py:40]: I would love to improve code quality before merging this. Otherwise we will forget to do this.

**Comment by haesleinhuepf** [comment]
> Would you mind to review it again

Please finish the code quality improvement (docstrings) first.

And have a great weekend btw!

File: scripts/data_normalizer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (SeverusYixin) [reply scripts/data_normalizer.py:40]: Got it

File: .github/workflows/book.yml (hunk 4 above)
Comment (SeverusYixin) [reply .github/workflows/book.yml]: right

**Comment by SeverusYixin** [comment]
> > Would you mind to review it again
> 
> Please finish the code quality improvement (docstrings) first.
> 
> And have a great weekend btw!

Same to you !

**Comment by SeverusYixin** [comment]
Hi@haesleinhuepf, me again :), when you have time, could you review if the code comments in this updated version are the type that you need!

**Review by haesleinhuepf — APPROVED** [review]
This looks very good @SeverusYixin , I'm going to merge it. Let's see what is the impace on the website.

**Comment by SeverusYixin** [comment]
I'm happy to hear that :)

**Pull request merged**